Vehicle rental error handling

evansiroky opened the pull request:
This PR mainly adds code to handle errors when fetching GBFS feeds. It classifies various errors by severity and will then display these errors in the plan API endpoint and also the vehicle rental API endpoint. Another thing this PR does is actually parse the system information data and also include that in the previously mentioned API endpoints. And since we're now tracking errors and handling them, this code also implements a TTL functionality that will automatically remove rental vehicles after they have outlived a user-defined TTL. This code is somewhat incomplete as it is only implemented fully for vehicle rentals (e-scooters). If this approach seems OK, it should also be implemented for car rentals and bike rentals to be consistent.

evansiroky committed:
Implement TTL for bike, car and vehicle rentals

evansiroky committed:
refactor: add detailed error handling through vehicle rental code

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/api/model/RentalInfo.java:
```diff
@@ -0,0 +1,29 @@
+    public final List<RentalUpdaterError> errors;
+
+    /**
+     * The system information about the vehicle rental company. This is take directly from the company's GBFS system
```
Typo: 'taken'

evansiroky replied:
fixed in 0b5b987b035c15264f45fc1eaf1e0cc5979fbe38

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/api/resource/GraphPathToTripPlanConverter.java:
```diff
@@ -752,6 +760,37 @@ private static void calculateElevations(Itinerary itinerary, Edge[] edges) {
+            Map<String, SystemInformation.SystemInformationData> systemInformationDataByNetwrok = vehicleRentalStationService
+                .getSystemInformationDataByNetwork();
+            for (String vehicleRentalNetwork : vehicleRentalNetworks) {
+                rentalInfo.put(vehicleRentalNetwork, new RentalInfo(errorsByNetwork.get(vehicleRentalNetwork),
```
Would it make sense to solely use `errorsByNetwork` to build `rentalInfo`?
For instance
```java
rentalInfo = errorsByNetwork.stream().map(...)...
```

evansiroky replied:
This comment may no longer relevant given the refactor in e67eacf0bcf74eb8378d2009d89c421ec29622c1

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/routing/vehicle_rental/RentalStation.java:
```diff
@@ -35,5 +35,31 @@
     public Set<String> networks = null;
+
+    /**
+     * The last time (in epoch seconds) that this rental station had updated information. This source of this
```
_*The*_ source of this info...

evansiroky replied:
fixed in 0b5b987b035c15264f45fc1eaf1e0cc5979fbe38

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/routing/vehicle_rental/RentalStation.java:
```diff
@@ -35,5 +35,31 @@
+    /**
+     * The last time (in epoch seconds) that this rental station had updated information. This source of this
+     * information varies according to the following fall-back methods:
+     *  - First try to use the value for a specific station or floating vehicle was updated according to the provider if
```
This looks like two sentences mashed around 'was updated'... please rephrase.

evansiroky replied:
fixed in 0b5b987b035c15264f45fc1eaf1e0cc5979fbe38

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/routing/vehicle_rental/RentalStation.java:
```diff
@@ -35,5 +35,31 @@
+     *  free_bike_status#data#bikes#bike#last_reported is available in feeds compliant with GBFS-2.1-RC+.
+     *  - If per-station or per-vehicle information is not available, try to use the feed-wide last update time. In the
+     *  context of the GBFS, this will be either the station_status#last_udated or free_bike_status#last_updated field.
+     *  - If feed-wide data on when the last report was, then use the current timestamp when the data was received by
```
First part of sentence is incomplete.

evansiroky replied:
fixed in 0b5b987b035c15264f45fc1eaf1e0cc5979fbe38

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/updater/vehicle_rental/VehicleRentalDataSource.java:
```diff
@@ -23,12 +25,10 @@
-    /** Returns true if there might have been changes to the stations */
-    boolean stationsUpdated();
+    /**
+     * @return a List of all errors that occurred during the most recent update.
```
Lower case list, or use the link notation (also same for `getRegions`).

evansiroky replied:
fixed in 0b5b987b035c15264f45fc1eaf1e0cc5979fbe38

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/updater/vehicle_rental/VehicleRentalUpdater.java:
```diff
@@ -64,30 +64,25 @@
 
         // Create graph writer runnable to apply these stations and regions to the graph
-        updaterManager.execute(new VehicleRentalGraphWriterRunnable(stations, regions));
+        updaterManager.execute(new VehicleRentalGraphWriterRunnable(
```
Can this constructor just use `source` as an argument?

evansiroky replied:
fixed in e67eacf0bcf74eb8378d2009d89c421ec29622c1

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/updater/vehicle_rental/VehicleRentalUpdater.java:
```diff
@@ -133,63 +129,121 @@ public void setup(Graph graph) {
+
+            // Apply stations to graph if a feed-wide error did not occur
+            if (!feedWideError) {
+                /* add any new stations that have fresh-enough data and update vehicle counts for existing stations */
```
Change to `//` comment.

evansiroky replied:
fixed in 0b5b987b035c15264f45fc1eaf1e0cc5979fbe38

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/util/DateUtils.java:
```diff
@@ -282,4 +282,11 @@ public static long absoluteTimeout(double relativeTimeoutSeconds) {
     }
+
+    /**
+     * Returns false if the given epochSeconds is more than ttlMinutes in the past.
```
Spell out time-to-live minutes.

evansiroky replied:
fixed in 0b5b987b035c15264f45fc1eaf1e0cc5979fbe38

binh-dam-ibigroup reviewed (approved):
Conditional approval on correcting the various typos and incomplete sentences as noted.
Also see the suggestion regarding building the map of errors by network.

br648 commented on src/main/java/org/opentripplanner/updater/bike_rental/BikeRentalUpdater.java:
```diff
@@ -166,8 +170,12 @@ public void run(Graph graph) {
             Set<BikeRentalStation> stationSet = new HashSet<>();
             Set<String> defaultNetworks = new HashSet<>(Arrays.asList(network));
             LOG.info("Updating {} rental bike stations.", stations.size());
-            /* add any new stations and update bike counts for existing stations */
```
Change comment to // as noted by Binh else where?

evansiroky replied:
fixed in 0b5b987b035c15264f45fc1eaf1e0cc5979fbe38

br648 reviewed (approved):
LGTM, One very minor comment.

evansiroky commented:
I'm assigning back to myself to add error code for bike and car rentals.

evansiroky committed:
refactor: copy pasta vehicle rental error handling to bike and car updaters

evansiroky committed:
refactor: implement various suggestions in PR review

See https://github.com/ibi-group/OpenTripPlanner/pull/53

evansiroky commented:
Ready for review again, this time with error handling fully implemented for all bike and car rental updaters too.

br648 reviewed (approved):
LGTM

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/api/resource/GraphPathToTripPlanConverter.java:
```diff
@@ -761,34 +763,49 @@ private static void calculateElevations(Itinerary itinerary, Edge[] edges) {
-            }
+        if (bikeRentalStationService != null && options.allowBikeRental) {
+            plan.bikeRentalInfo = makeRentalInfo(
+                bikeRentalStationService.getErrorsByNetwork(),
+                bikeRentalStationService.getSystemInformationDataByNetwork()
```
Potential refactoring: move `getErrorsByNetwork`, `getSystemInformationDataByNetwork` to a common type/interface implemented by `bikeRentalStationService`, `carRentalStationService`, `vehicleRentalService`, and pass that type to `makeRentalInfo()`. Even better, move `makeRentalInfo` to that common type.

evansiroky replied:
If I try to refactor and combine this logic, I might as well combine all of the vehicle rental logic which would take some time. That was not advised prior to starting this work.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/updater/car_rental/Car2GoCarRentalDataSource.java:
```diff
@@ -5,6 +5,7 @@
 import org.opentripplanner.routing.car_rental.CarFuelType;
 import org.opentripplanner.routing.car_rental.CarRentalStation;
 import org.opentripplanner.routing.graph.Graph;
+import org.opentripplanner.updater.RentalUpdaterError;
```
This import does not appear to be used.

evansiroky replied:
fixed in a78c584d8f78dba18d9033dbd56150a0cb695f88

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/updater/car_rental/Car2GoCarRentalDataSource.java:
```diff
@@ -14,6 +15,7 @@
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.HashSet;
+import java.util.List;
```
This import does not appear to be used.

evansiroky replied:
fixed in a78c584d8f78dba18d9033dbd56150a0cb695f88

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/updater/bike_rental/BikeRentalUpdater.java:
```diff
@@ -141,76 +144,122 @@ public void setup(Graph graph) throws InterruptedException, ExecutionException {
-            /* remove existing stations that were not present in the update */
-            List<BikeRentalStation> toRemove = new ArrayList<BikeRentalStation>();
+
+            // remove existing stations that were not present in the update
```
The block below is similar to the one in `CarRentalUpdater` and could be extracted to a method in `PollingGraphUpdater`

evansiroky replied:
If I try to refactor and combine this logic, I might as well combine all of the vehicle rental logic which would take some time. That was not advised prior to starting this work.

binh-dam-ibigroup reviewed (changes requested):
Please consider refactoring `bikeRentalStationService`, `carRentalStationService`, `vehicleRentalService` as indicated, and also consider refactoring updating station lists in `CarRentalUpdater`, `BikeRentalUpdater`

evansiroky committed:
refactor: remove unused imports in various files

binh-dam-ibigroup reviewed (approved):
I can live with the current changes.

landonreed commented on src/main/java/org/opentripplanner/routing/vehicle_rental/RentalStation.java:
```diff
@@ -35,5 +35,31 @@
+     *  - If feed-wide data on the last reported timestamp is not available, then use the current timestamp when the
+     *  data was received by OTP.
+     */
+    @XmlAttribute
```
Should we clear out these XmlAttribute annotations?

evansiroky replied:
Seems out-of-scope for this PR.

landonreed commented on src/main/java/org/opentripplanner/updater/bike_rental/GenericJsonBikeRentalDataSource.java:
```diff
@@ -41,10 +49,8 @@
-        jsonParsePath = jsonPath;
-        headerName = "Default";
-        headerValue = null;
+    public GenericJsonBikeRentalDataSource(RentalUpdaterError.Severity severityFailureType, String jsonPath) {
```
It's somewhat unclear why this constructor accepts an error type. Maybe the javadoc could be updated.

evansiroky replied:
Added in ab194e0

landonreed commented on src/main/java/org/opentripplanner/updater/bike_rental/GenericJsonBikeRentalDataSource.java:
```diff
@@ -26,12 +28,18 @@
 
     private static final Logger log = LoggerFactory.getLogger(GenericJsonBikeRentalDataSource.class);
+
+    private final RentalUpdaterError.Severity severityFailureType;
```
Might be good to add javadoc for this new field (its purpose is not as inherently obvious as the others).

evansiroky replied:
Added in ab194e0

landonreed commented on src/main/java/org/opentripplanner/updater/bike_rental/GenericJsonBikeRentalDataSource.java:
```diff
@@ -84,20 +98,26 @@ public boolean update() {
+                data.close();
+            } catch (IOException e) {
+                log.warn("An error occurred while closing data stream {}");
+                e.printStackTrace();
```
This log message has a slot for a param that should be removed. Also, I think you can just move the exception as the second arg to the warn method and it will print the stack trace tied to the warning.

evansiroky replied:
Refactored in ab194e0

landonreed commented on src/main/java/org/opentripplanner/updater/bike_rental/GenericJsonBikeRentalDataSource.java:
```diff
@@ -84,20 +98,26 @@ public boolean update() {
             return false;
         } catch (IOException e) {
-            log.warn("Error reading bike rental feed from " + url, e);
+            addError("Error reading bike rental feed from " + url);
```
I'm not crazy about removing the logging of all of these exceptions. These are important clues to debugging when reading the logs. Perhaps `addError` should have an arg for exception and log it just as it was being logged here (maybe with a null check thrown in for good measure).

evansiroky replied:
Added in ab194e0

landonreed reviewed (changes requested):
Looks good, just a few comments.

evansiroky committed:
Refactor rental error handling

- Move error response to main response entity
- Improve error handling
- Generate a feed-wide error if a feed has both ALL_STATIONS and ALL_FLOATING_VEHICLES errors

evansiroky committed:
Address various comments in PR review

See https://github.com/ibi-group/OpenTripPlanner/pull/53

evansiroky commented:
I just refactored a few items. Namely, I moved the rental information up to the root level in the Response to trip plans. This was needed because sometimes a path wouldn't be calculated and therefore, no rental info would be added to a plan because no plan would exist. Also, this code now generates a feed-wide error if both stations and vehicles couldn't be fetched.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/updater/bike_rental/BikeRentalUpdater.java:
```diff
@@ -54,6 +54,8 @@
 
     private int timeToLiveMinutes;
 
+    private String networkName = "default";
```
What is the difference with private field `network` above?

evansiroky replied:
I also had to do a long double take when I saw this too. The `network` field could be thought of as `compatibleNetworks` as in a comma-separated list of strings that identify the compatible networks where a vehicle can be dropped off at. The `networkName` here is the name of the specific network of this particular bike rental updater. I will strongly advocate for rewording of this when updating to GBFS 2.2 and even a check on whether this is even needed anymore.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/updater/bike_rental/GbfsBikeRentalDataSource.java:
```diff
@@ -64,12 +64,19 @@ public boolean update() {
-        boolean updatesFound = stationInformationSource.update();
-        updatesFound |= stationStatusSource.update();
+        boolean stationInfoFound = stationInformationSource.update();
+        stationInfoFound &= stationStatusSource.update();
```
I would use `&&` instead of `&=`.

evansiroky replied:
Refactored in 19d09a6

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/updater/bike_rental/GenericJsonBikeRentalDataSource.java:
```diff
@@ -45,6 +48,9 @@
     /**
      * Construct superclass
      *
+     * @param severityFailureType The severity level to create an error with when fetching from this particular data
```
Suggestion: "The security level *given to an error* when...".

evansiroky replied:
Refactored in 19d09a6

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/updater/bike_rental/GenericJsonBikeRentalDataSource.java:
```diff
@@ -45,6 +48,9 @@
      * Construct superclass
      *
+     * @param severityFailureType The severity level to create an error with when fetching from this particular data
+     *                            source. This can be used to differentiate different error types for fetching from
```
Suggestion: "This can be used to *define* different error types *when* fetching..."

evansiroky replied:
Refactored in 19d09a6

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/updater/bike_rental/GenericJsonBikeRentalDataSource.java:
```diff
@@ -55,6 +61,9 @@ public GenericJsonBikeRentalDataSource(RentalUpdaterError.Severity severityFailu
 
     /**
      *
+     * @param severityFailureType The severity level to create an error with when fetching from this particular data
```
Suggestion: "The security level *given to an error* when...".

evansiroky replied:
Refactored in 19d09a6

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/updater/bike_rental/GenericJsonBikeRentalDataSource.java:
```diff
@@ -55,6 +61,9 @@ public GenericJsonBikeRentalDataSource(RentalUpdaterError.Severity severityFailu
     /**
      *
+     * @param severityFailureType The severity level to create an error with when fetching from this particular data
+     *                            source. This can be used to differentiate different error types for fetching from
```
Suggestion: "This can be used to *define* different error types *when* fetching..."

evansiroky replied:
Refactored in 19d09a6

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/updater/vehicle_rental/GenericGbfsService.java:
```diff
@@ -231,6 +238,7 @@ public void update () {
                     RentalUpdaterError.Severity.FEED_WIDE,
+                    e,
                     "Failed to deserialize gbfs.json response: %s",
                     e
```
Need a comment to explain why `e` is passed twice.

evansiroky replied:
I changed this to be `e.toString()` so it more intuitively matches the string template argument type in 19d09a6.

binh-dam-ibigroup reviewed (approved):
Conditional approval, please advise on duplicate `network`/`networkName` field and address formatting suggestions I made.

br648 commented on src/main/java/org/opentripplanner/updater/bike_rental/GbfsBikeRentalDataSource.java:
```diff
@@ -56,14 +61,38 @@ public void setBaseUrl (String url) {
+     * Adds an error message to the list of errors and also logs the error message.
+     */
+    private void addError(RentalUpdaterError.Severity severity, String message) {
+        message = String.format("%s (feed: %s)", message, networkName);
```
Should "feed:" be network name?

evansiroky replied:
Renamed in 19d09a6

br648 commented on src/main/java/org/opentripplanner/updater/bike_rental/GenericJsonBikeRentalDataSource.java:
```diff
@@ -84,20 +114,27 @@ public boolean update() {
             return false;
         } catch (JsonProcessingException e) {
-            log.warn("Error parsing bike rental feed from " + url + "(bad JSON of some sort)", e);
+            addError("Error parsing bike rental feed from " + url + "(bad JSON of some sort)", e);
```
Should this use String.format instead?

evansiroky replied:
It didn't before and I hope this is refactored away anyways in the GBFS 2.2 update, so it doesn't seem worth the effort.

br648 commented on src/main/java/org/opentripplanner/updater/vehicle_rental/GenericGbfsService.java:
```diff
@@ -148,24 +166,42 @@ private void setRegionsFromConfig(JsonNode config) {
+     * Adds an error message to the list of errors and also logs the error message.
+     */
+    private void addError(RentalUpdaterError.Severity severity, Exception e, String message) {
+        message = String.format("%s (feed: %s)", message, networkName);
```
Should "feed:" be network name?

evansiroky replied:
Renamed in 19d09a6

br648 reviewed (approved):
some minor comments.

evansiroky committed:
Address various PR review comments.

See https://github.com/ibi-group/OpenTripPlanner/pull/53
